sdk.go

```diff
@@ -96,27 +121,240 @@ const (
+// an unsafe.Pointer. Manually writing inside the C array might break the
+// internal logic of sdk.PluginEvents thus leading to non-deterministic
+// behavior.
+//
+// This is intended to be used as a slab memory allocator. PluginEvents
+// are supposed to be stored inside the plugin instance state to avoid
+// useless reallocations, and should be used to create plugin events and
+// write data in them. Unlike slices, the events contained in the list
+// can only be accessed by using the Get and Len methods to enforce safe
+// memory accesses. Ideally, the list is meant to be large enough to contain
+// the maximum number of events that the plugin is capable of producing with
+// plugin_next_batch. The plugin_next symbol should only work on the first
+// event of the list instead.
+//
+// The underlying C memory managed by this interface is out of the scope of
+// garbage collection, and memory must be manually deallocated by invoking
+// the Free method. Using instances of PluginEvents after invoking its Free()
+// method might lead to non-deterministic behavior.
+//
+// Here is an example of usage:
+//	func plugin_open(pState unsafe.Pointer, params *C.char, rc *int32) unsafe.Pointer {
+//		...
+//		// Create instance of sdk.PluginEvents
+//		pluginEvents, err := sdk.NewPluginEvents(maxNextBatchEvents, int64(sdk.MaxEvtSize))
+//		if err != nil {
+//			*rc = sdk.SSPluginFailure
+//			return nil
+//		}
+//
+//		// Store pluginEvents inside the plugin instance state
+//		is := &instanceState{
+//			...
+//			pluginEvents:       pluginEvents,
+//		}
+//		handle := state.NewStateContainer()
+//		state.SetContext(handle, unsafe.Pointer(is))
+//		*rc = sdk.SSPluginSuccess
+//		return handle
+//	}
+//
+//	func plugin_next(pState unsafe.Pointer, iState unsafe.Pointer, retEvt **C.ss_plugin_event) int32 {
+//		// Grab an instance of the event through sdk.PluginEvents
+//		is := (*instanceState)(state.Context(iState))
+//		event := is.pluginEvents.Get(0)
+//
+//		// Write the event data and set the timestamp
+//		eventData := "Sample event data... Hello World!"
+//		eventWriter := event.Writer()
+//		eventWriter.Write([]byte(eventData))
+//		event.SetTimestamp(uint64(time.Now().Unix()) * 1000000000)
+//
+//		// Set the result pointer to the internal event array buffer
+//		*retEvt = (*C.ss_plugin_event)(is.pluginEvents.ArrayPtr())
+//		return sdk.SSPluginSuccess
+//	}
+//
+type PluginEvents interface {
+	// Get returns an instance of sdk.PluginEvent at the eventIndex
+	// position inside the list.
+	Get(eventIndex int) PluginEvent
+	//
+	// Len returns the size of the list, namely the number of events
+	// it contains. Using Len coupled with Get allows iterating over
+	// all the events of the list.
+	Len() int
+	//
+	// Free takes care of de-allocating all the memory managed by
+	// instances of PluginEvents. Note that using the same instance
+	// after invoking its Free method might lead to non-determinitic
+	// behavior.
+	Free()
```

Comment:
To avoid exposing Free(), we might use [runtime.SetFinalizer()](https://pkg.go.dev/runtime#SetFinalizer) instead.
Wdyt?

Reply:
That sounds like a great idea!